File: controllers/book_controllers.py
from flask import jsonify, request
from models.book_model import books

# POST - Create book
def create_book():
    data = request.get_json()

    if not all(key in data for key in ("title", "author", "year", "genre")):
        return jsonify({"message": "All fields are required: title, author, year, genre."}), 400

    new_book = {
        "id": len(books) + 1,
        "title": data["title"],
        "author": data["author"],
        "year": data["year"],
        "genre": data["genre"]
    }

    books.append(new_book)
    return jsonify({"message": "Book created successfully!", "book": new_book}), 201


# GET - List all books (with optional author filter)
def list_books():
    author = request.args.get("author")

    if author:
        filtered = [b for b in books if author.lower() in b["author"].lower()]
        return jsonify(filtered)

    return jsonify(books)


# PUT - Update book
def update_book(id):
    data = request.get_json()
    book = next((b for b in books if b["id"] == id), None)

    if not book:
        return jsonify({"message": "Book not found."}), 404

    book["title"] = data.get("title", book["title"])
    book["author"] = data.get("author", book["author"])
    book["year"] = data.get("year", book["year"])
    book["genre"] = data.get("genre", book["genre"])

    return jsonify({"message": "Book updated successfully!", "book": book})


# DELETE - Delete book
def delete_book(id):
    global books
    book = next((b for b in books if b["id"] == id), None)

    if not book:
        return jsonify({"message": "Book not found."}), 404

    books = [b for b in books if b["id"] != id]
    return jsonify({"message": "Book deleted successfully!"})

File: controllers/book_controllers.js
from const Book = require('../models/bookModel');

exports.createBook = async (req, res) => {
    try {
        const { title, author, year } = req.body;
        const newBook = new Book({ title, author, year });
        await newBook.save();
        res.status(201).json({ message: 'Livro criado com sucesso!', book: newBook });
    } catch (error) {
        res.status(500).json({ message: 'Erro ao criar livro', error: error.message });
    }
};

exports.getAllBooks = async (req, res) => {
    try {
        const books = await Book.find();
        res.status(200).json({ books });
    } catch (error) {
        res.status(500).json({ message: 'Erro ao buscar livros', error: error.message });
    }
};


exports.updateBook = async (req, res) => {
    try {
        const { id } = req.params;
        const { title, author, year } = req.body;
        const updatedBook = await Book.findByIdAndUpdate(id, { title, author, year }, { new: true });
        
        if (!updatedBook) {
            return res.status(404).json({ message: 'Livro não encontrado' });
        }

        res.status(200).json({ message: 'Livro atualizado com sucesso', book: updatedBook });
    } catch (error) {
        res.status(500).json({ message: 'Erro ao atualizar livro', error: error.message });
    }
};

exports.deleteBook = async (req, res) => {
    try {
        const { id } = req.params;
        const deletedBook = await Book.findByIdAndDelete(id);

        if (!deletedBook) {
            return res.status(404).json({ message: 'Livro não encontrado' });
        }

        res.status(200).json({ message: 'Livro excluído com sucesso', book: deletedBook });
    } catch (error) {
        res.status(500).json({ message: 'Erro ao excluir livro', error: error.message });
    }
};
